### ISSUE TITLE
Unable to monitor HikariCP pools created with play-slick due to conflicting poolName

### ISSUE THREAD

**Issue by balagez** [playframework/play-slick#348]
There's a [bug in Slick's HikariCP code](https://github.com/slick/slick/issues/1326) which makes Slick use the `name` parameter of `forConfig` as the pool name. If an application that uses play-slick has more than one connection pools, because of this Slick bug, all pools are created with the same `db` pool name. This is serious problem if we want to [monitor HikariCP connection pools](https://github.com/brettwooldridge/HikariCP/wiki/MBean-%28JMX%29-Monitoring-and-Management) because HikariCP registers MBeans using the `poolName` and these names are conflicting, so only the first MBean is registered successfully, the rest of them fail.

Of course the ultimate solution would be to fix the Slick bug, but until then, maybe play-slick could do something to provide Slick with the correct pool names.

Example play-slick configuration based on the [Play framework documentation](https://www.playframework.com/documentation/2.4.x/PlaySlick#Database-Configuration):

```
slick.dbs.read.driver = "slick.driver.MySQLDriver$"
slick.dbs.read.db.driver = "com.mysql.jdbc.Driver"
slick.dbs.read.db.url = "jdbc:mysql://localhost:3306/dbname"
slick.dbs.read.db.poolName = "read"
slick.dbs.read.db.registerMbeans = true

slick.dbs.write.driver = "slick.driver.MySQLDriver$"
slick.dbs.read.db.driver = "com.mysql.jdbc.Driver"
slick.dbs.read.db.url = "jdbc:mysql://localhost:3306/dbname"
slick.dbs.read.db.poolName = "write"
slick.dbs.read.db.registerMbeans = true
```

And relevant parts of the HikariCP debug log:

```
2016-02-10 18:51:03,846 [DEBUG] com.zaxxer.hikari.HikariConfig - HikariCP pool db configuration:
...
2016-02-10 18:51:03,859 [DEBUG] com.zaxxer.hikari.HikariConfig - poolName........................db
2016-02-10 18:51:03,859 [DEBUG] com.zaxxer.hikari.HikariConfig - readOnly........................false
2016-02-10 18:51:03,859 [DEBUG] com.zaxxer.hikari.HikariConfig - registerMbeans..................true
...
2016-02-10 18:51:03,862 [INFO] com.zaxxer.hikari.HikariDataSource - HikariCP pool db is starting.
2016-02-10 18:51:03,912 [DEBUG] com.zaxxer.hikari.HikariConfig - HikariCP pool db configuration:
...
2016-02-10 18:51:03,921 [DEBUG] com.zaxxer.hikari.HikariConfig - poolName........................db
2016-02-10 18:51:03,921 [DEBUG] com.zaxxer.hikari.HikariConfig - readOnly........................false
2016-02-10 18:51:03,921 [DEBUG] com.zaxxer.hikari.HikariConfig - registerMbeans..................true
...
2016-02-10 18:51:03,922 [INFO] com.zaxxer.hikari.HikariDataSource - HikariCP pool db is starting.
2016-02-10 18:51:03,923 [ERROR] com.zaxxer.hikari.pool.HikariMBeanElf - You cannot use the same pool name for separate pool instances.
```

**Comment by dotta** [comment]
@balagez Thanks for the pointer. If I'm understanding things correctly, we should document this problem and suggest to declare the `slick.dbs.<db-name>.db.poolName` as a workaround until a Slick release that contains the fix is available.

**Comment by balagez** [comment]
@dotta Unfortunately there's no workaround for end users because due to that bug I mentioned Slick currently doesn't read `poolName` from configuration but [blindly uses the config path as the pool name](https://github.com/slick/slick/blob/3.1.0/slick-hikaricp/src/main/scala/slick/jdbc/hikaricp/HikariCPJdbcDataSource.scala#L50), which in case of play-slick will always be `db` since play-slick [calls `DatabaseConfig.forConfig` with `path = ""`](https://github.com/playframework/play-slick/blob/master/src/core/src/main/scala/play/api/db/slick/SlickApi.scala#L89) and [Slick appends the constant `db` suffix](https://github.com/slick/slick/blob/3.1.0/slick/src/main/scala/slick/backend/DatabaseConfig.scala#L73). But simply changing `path` doesn't work either because then Slick won't find the configuration as the contents of `config` and the config path get out of sync.

The only opportunity for a workaround I see is rather clumsy, that is to call `DatabaseConfig.forConfig` with an artificially "padded" config object where all keys in that config are prefixed with the `<db-name>` and then setting `path` to `<db-name>` too. And I'm not at all sure that `com.typesafe.config.Config` provides a way to prefix all keys so this might not even be a viable option.

In the meantime [I patched Slick](https://github.com/slick/slick/pull/1439) so the fix is coming in Slick 3.2.0. It's up to you whether you want to implement such a clumsy workaround or just wait for the next Slick release. In the latter case I'm okay with closing this issue, and maybe the documentation could mention that currently it's impossible to implement JMX monitoring with play-slick/Slick 3.1.0.

**Comment by marcospereira** [comment]
Related to #344.

@balagez, your patch is part of slick 3.2.0-RC1 version which is being used by play-slick. So I'm closing this.
